azurelinuxagent/common/protocol/wire.py

```diff
@@ -913,49 +913,51 @@ def _call_hostplugin_with_container_check(self, host_func):
-            try:
-                ret = direct_func()
+    def __send_request_using_direct_channel(self, direct_func):
+        ret = None
+        try:
+            ret = direct_func()
 
-                # Different direct channel functions report failure in different ways: by returning None, False,
-                # or raising ResourceGone or InvalidContainer exceptions.
-                if not ret:
-                    logger.periodic_info(logger.EVERY_HOUR, "[PERIODIC] Request failed using the direct channel, "
-                                                            "switching to host plugin.")
-            except (ResourceGoneError, InvalidContainerError) as e: # pylint: disable=C0103
+            # Different direct channel functions report failure in different ways: by returning None, False,
+            # or raising ResourceGone or InvalidContainer exceptions.
+            if not ret:
                 logger.periodic_info(logger.EVERY_HOUR, "[PERIODIC] Request failed using the direct channel, "
-                                                        "switching to host plugin. Error: {0}".format(ustr(e)))
+                                                        "trying host plugin next.")
+        except (ResourceGoneError, InvalidContainerError) as e:  # pylint: disable=C0103
+            logger.periodic_info(logger.EVERY_HOUR, "[PERIODIC] Request failed using the direct channel, "
+                                                    "trying host plugin next. Error: {0}".format(ustr(e)))
 
-            if ret:
-                return ret
-        else:
-            logger.periodic_info(logger.EVERY_HALF_DAY, "[PERIODIC] Using host plugin as default channel.")
+        if ret and HostPluginProtocol.is_default_channel():
+            # Set direct channel as default if it succeeded.
+            logger.info("Switching default channel to WireServer.")
+            HostPluginProtocol.set_default_channel(False)
 
-        ret = self._call_hostplugin_with_container_check(host_func)
+        return ret
 
-        if not HostPluginProtocol.is_default_channel():
-            logger.info("Setting host plugin as default channel from now on. "
-                        "Restart the agent to reset the default channel.")
-            HostPluginProtocol.set_default_channel(True)
+    def send_request_using_appropriate_channel(self, direct_func, host_func):
+        primary_ret = None
+        try:
+            primary_ret = self.__send_request_using_host_channel(host_func) if HostPluginProtocol.is_default_channel() \
+                else self.__send_request_using_direct_channel(direct_func)
+        except Exception:
+            pass
 
-        return ret
+        if primary_ret:
+            return primary_ret
+
+        secondary_ret = self.__send_request_using_direct_channel(direct_func) if HostPluginProtocol.is_default_channel() \
+            else self.__send_request_using_host_channel(host_func)
+        return secondary_ret
```

Comment:
it is also confusing that we use try/except when using the primary channel, but not when using the secondary channel. when both the primary and secondary channels could be either direct or the host ga plugin

Reply:
+1

Reply:
Why is that confusing? We want to retry if primary fails (with secondary), but if secondary fails, we give up and report failure.

Reply:
it may be a consequence of the code above... once that is clarified this may become clearer too

Reply:
As discussed offline, I'll rewrite this method to be clearer.